snafu-derive/src/lib.rs

```diff
@@ -903,45 +938,53 @@ impl syn::parse::Parse for Backtrace {
     }
 }
 
+/// A SnafuAttribute represents one SNAFU-specific attribute inside of `#[snafu(...)]`.  For
+/// example, in `#[snafu(visibility(pub), display("hi"))]`, `visibility(pub)` and `display("hi")`
+/// are each a SnafuAttribute.
+///
+/// We store the location in the source where we found the attribute (as a `TokenStream`) along
+/// with the data.  The location can be used to give accurate error messages in case there was a
+/// problem with the use of the attribute.
 enum SnafuAttribute {
-    Display(UserInput),
-    Visibility(UserInput),
-    Source(Vec<Source>),
-    Backtrace(Backtrace),
-    DocComment(String),
+    Display(proc_macro2::TokenStream, UserInput),
+    Visibility(proc_macro2::TokenStream, UserInput),
+    Source(proc_macro2::TokenStream, Vec<Source>),
+    Backtrace(proc_macro2::TokenStream, Backtrace),
+    DocComment(proc_macro2::TokenStream, String),
 }
 
 impl syn::parse::Parse for SnafuAttribute {
     fn parse(input: syn::parse::ParseStream) -> SynResult<Self> {
         use syn::{Expr, Ident, Visibility};
 
+        let input_tts = input.cursor().token_stream();
```

Comment:
Huh. I saw these functions in syn, but didn't realize that they could be used this way. Cool!